### ISSUE TITLE
[Bug]: Dynamic loading LoRA is not working properly

### ISSUE THREAD

**Issue by DonghoKang** [vllm-project/vllm#18372]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

export VLLM_ALLOW_RUNTIME_LORA_UPDATING=True
</details>

### 🐛 Describe the bug

I'm using vllm 0.6.5 version now.
And, I make a script for unloading and loading lora weights dynamically while vllm serving.

For example
curl -X POST "http://localhost:8012/v1/unload_lora_adapter" -H "Content-Type":"application/json" --data '{ "lora_name":"ex" }'
curl -X POST "http://localhost:8012/v1/load_lora_adapter" -H "Content-Type":"application/json" --data '{ "lora_name":"ex", "lora_path":"${LORA_PATH}"}'

When I run this script, it return success always even I pass a path as a "LORA_PATH" which is not exist.
And model output does not affected.
This is server log.

INFO:     ::1:43228 - "POST /v1/load_lora_adapter HTTP/1.1" 200 OK

But, if I use `--lora-modules` explicitly when I run vllm serve, it works fine.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by jeejeelee** [comment]
Have you tried the latest version

**Comment by Fluchw** [comment]
> Have you tried the latest version你试过最新版本吗

I'm the same way

```bash
vllm-apiu@ununtu:~/projects/vllm$ uv pip show vllm
Name: vllm
Version: 0.9.1
Location: /home/u/projects/vllm/.venv/lib/python3.12/site-packages
Requires: aiohttp, blake3, cachetools, cloudpickle, compressed-tensors, depyf, einops, fastapi, filelock, gguf, huggingface-hub, lark, llguidance, lm-format-enforcer, mistral-common, msgspec, ninja, numba, numpy, openai, opencv-python-headless, opentelemetry-api, opentelemetry-exporter-otlp, opentelemetry-sdk, opentelemetry-semantic-conventions-ai, outlines, partial-json-parser, pillow, prometheus-client, prometheus-fastapi-instrumentator, protobuf, psutil, py-cpuinfo, pydantic, python-json-logger, pyyaml, pyzmq, ray, regex, requests, scipy, sentencepiece, setuptools, six, tiktoken, tokenizers, torch, torchaudio, torchvision, tqdm, transformers, typing-extensions, watchfiles, xformers, xgrammar
Required-by:
```

```bash
curl -X 'POST' \
  'http://127.0.0.1:6379/v1/chat/completions' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  
  "messages": [
    {
      "content": "What are you doing",
      "role": "user"
    }]
}'
```

```bash
{
  "id": "chatcmpl-eae8f691241448e091950dcf1177c5d9",
  "object": "chat.completion",
  "created": 1750909345,
  "model": "/home/u/projects/llm_mod/gemma-3-12b-it",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "reasoning_content": null,
        "content": "As a large language model, I don't \"do\" things in the way a person does. Right now, I'm responding to your question! \n\nSpecifically, I'm processing your input (\"What are you doing\"), understanding what you're asking, and generating a relevant and informative response based on my training data. \n\nThink of it like this: I'm constantly waiting for a prompt, and when I get one, I use my knowledge and algorithms to produce text. 😊\n\n\n\nIs there anything else I can help you with?",
        "tool_calls": []
      },
      "logprobs": null,
      "finish_reason": "stop",
      "stop_reason": 106
    }
  ],
  "usage": {
    "prompt_tokens": 13,
    "total_tokens": 127,
    "completion_tokens": 114,
    "prompt_tokens_details": null
  },
  "prompt_logprobs": null,
  "kv_transfer_params": null
}
```
When I activate Lora
```bash
curl -X 'POST' \
  'http://127.0.0.1:6379/v1/chat/completions' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "model":"chat9",
  "messages": [
    {
      "content": "What are you doing",
      "role": "user"
    }]
}'
```
```bash
{
  "id": "chatcmpl-c5b1bc8182d04c38b683e9c38c117a0e",
  "object": "chat.completion",
  "created": 1750909411,
  "model": "chat9",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "reasoning_content": null,
        "content": "As a large language model, I don't \"do\" things in the way a person does. Right now, I'm responding to your question! \n\nMore specifically, I'm processing your input (\"What are you doing\"), understanding what you're asking, and generating a relevant and informative answer based on the vast amount of text data I've been trained on.\n\nThink of it like this: I'm running a complex computer program designed to understand and respond to language. 😊\n\n\n\nIs there anything else I can help you with?",
        "tool_calls": []
      },
      "logprobs": null,
      "finish_reason": "stop",
      "stop_reason": null
    }
  ],
  "usage": {
    "prompt_tokens": 13,
    "total_tokens": 127,
    "completion_tokens": 114,
    "prompt_tokens_details": null
  },
  "prompt_logprobs": null,
  "kv_transfer_params": null
}
```

**Comment by ojh31** [comment]
Same for me. I'm on Mac so I've pip installed the latest version from github (0.10.1.dev178+g0d0cc9e15). The following script shows that the first two lora requests seem to work as expected but then I am unable to update the lora adapter a third time.
```
import os
import subprocess
import time
import torch
import requests
from pathlib import Path
from typing import Dict, Any
from transformers import AutoTokenizer, AutoModelForCausalLM
from peft import LoraConfig, get_peft_model, TaskType


def start_vllm_server():
    """Start VLLM server as a subprocess."""
    cmd = [
        "python",
        "-m",
        "vllm.entrypoints.openai.api_server",
        "--model",
        "AlignmentResearch/Llama-3.3-Tiny-Instruct-boolq",
        "--enable-lora",
        "--max-lora-rank",
        "128",
    ]

    env = os.environ.copy()
    env["VLLM_ALLOW_RUNTIME_LORA_UPDATING"] = "True"

    process = subprocess.Popen(cmd, env=env)

    # Wait for server to start
    print("Starting VLLM server...")
    time.sleep(10)

    return process


def generate_with_lora(
    base_url: str, prompts: list[str], lora_request: Dict[str, Any]
) -> list[str]:
    """Generate responses using the VLLM server with LoRA adapter."""
    url = f"{base_url}/v1/completions"
    headers = {"Content-Type": "application/json"}

    responses = []
    for prompt in prompts:
        data = {
            "model": "AlignmentResearch/Llama-3.3-Tiny-Instruct-boolq",
            "prompt": prompt,
            "max_tokens": 5,
            "temperature": 0.0,
            "lora_request": lora_request,
        }

        response = requests.post(url, headers=headers, json=data)
        if response.status_code == 200:
            result = response.json()
            responses.append(result["choices"][0]["text"])
        else:
            print(f"Error: {response.status_code} - {response.text}")
            responses.append("")

    return responses


def format_chat_prompts(tokenizer, prompts: list[str]) -> list[str]:
    """Format prompts using chat template."""
    formatted = []
    for prompt in prompts:
        messages = [{"role": "user", "content": prompt}]
        formatted_prompt = tokenizer.apply_chat_template(
            messages, tokenize=False, add_generation_prompt=True
        )
        formatted.append(formatted_prompt)
    return formatted


def lora_weights_for_vllm(peft_model, tensor_shapes):
    """
    Extract LoRA weights in vLLM format, similar to GRPOTrainer.lora_weights_for_vllm()
    """
    lora_state_dict = {}
    for name, param in peft_model.named_parameters():
        if "policy" in name:  # Only policy parameters (like GRPOTrainer)
            lora_state_dict[name] = param.data.detach().cpu().clone()

    # Rename everything to match vLLM format (similar to GRPOTrainer)
    lora_weights = {}
    for key, value in lora_state_dict.items():
        new_key = key
        # Remove any module prefixes (similar to GRPOTrainer)
        if "_fsdp_wrapped_module." in new_key:
            new_key = new_key.replace("_fsdp_wrapped_module.", "")
        if "module." in new_key:
            new_key = new_key.replace("module.", "")

        # Get the shape from the original model and reshape
        if new_key in tensor_shapes:
            shape = tensor_shapes[new_key]
            # Remove "policy." prefix for vLLM format (like GRPOTrainer)
            final_key = new_key.replace("policy.", "")
            lora_weights[final_key] = value.reshape(shape)
        else:
            # If shape not found, use original value
            final_key = new_key.replace("policy.", "")
            lora_weights[final_key] = value

    return lora_weights


def save_adapter_checkpoint(peft_model, lora_config, tensor_shapes, lora_path_str):
    """Save adapter checkpoint similar to GRPOTrainer.save_adapter_checkpoint()"""
    lora_path = Path(lora_path_str)
    lora_path.mkdir(parents=True, exist_ok=True)

    # Get LoRA weights in vLLM format
    lora_weights = lora_weights_for_vllm(peft_model, tensor_shapes)

    # Save using torch.save and config.save_pretrained (like GRPOTrainer)
    torch.save(lora_weights, lora_path / "adapter_model.bin")
    lora_config.save_pretrained(str(lora_path))

    print(f"Saved adapter checkpoint to {lora_path}")


def main():
    # Start server
    server_process = start_vllm_server()

    try:
        base_url = "http://localhost:8000"
        model_name = "AlignmentResearch/Llama-3.3-Tiny-Instruct-boolq"

        # Create working directory
        work_dir = Path("./checkpoints/scripts_test_lora_adapter_update")
        work_dir.mkdir(parents=True, exist_ok=True)

        model = AutoModelForCausalLM.from_pretrained(
            model_name, torch_dtype=torch.bfloat16, device_map="auto"
        )
        tokenizer = AutoTokenizer.from_pretrained(model_name)

        # Test prompts
        test_prompts = [
            "Hello, how are you?",
            "What is the capital of France?",
            "How many fingers am I holding up?",
        ]

        # Format prompts using tokenizer
        prompts_formatted = format_chat_prompts(tokenizer, test_prompts)

        print("Formatted prompts:")
        for i, prompt in enumerate(prompts_formatted):
            print(f"{i}: {prompt}")

        # Create initial LoRA adapter and apply it to model
        lora_config = LoraConfig(
            task_type=TaskType.CAUSAL_LM,
            r=16,
            lora_alpha=32,
            target_modules="all-linear",
            lora_dropout=0.1,
        )
        peft_model = get_peft_model(model, lora_config, adapter_name="policy")

        for param in peft_model.parameters():
            if param.requires_grad:
                param.data = torch.zeros_like(param.data)

        # Store tensor shapes for later reshaping (like GRPOTrainer)
        tensor_shapes = {k: v.shape for k, v in peft_model.named_parameters()}

        # Save initial adapter using GRPOTrainer method
        lora_path_0 = work_dir / "lora_0"
        save_adapter_checkpoint(
            peft_model, lora_config, tensor_shapes, str(lora_path_0)
        )

        lora_request_0 = {
            "lora_int_id": 0,
            "lora_name": "policy_0",
            "lora_path": str(lora_path_0),
        }

        # Generate with initial adapter
        print("\nGenerating with initial adapter...")
        responses_0 = generate_with_lora(base_url, prompts_formatted, lora_request_0)

        print("lora_request_0 responses:")
        for i, response in enumerate(responses_0):
            print(f"{i}: {response}")

        # Modify LoRA parameters in-place
        print("\nModifying LoRA parameters in-place with large random values...")
        for param in peft_model.parameters():
            if param.requires_grad:
                param.data = torch.randn_like(param.data) * 1e6

        # Save modified adapter to new path using GRPOTrainer method
        lora_path_1 = work_dir / "lora_1"
        save_adapter_checkpoint(
            peft_model, lora_config, tensor_shapes, str(lora_path_1)
        )

        lora_request_1 = {
            "lora_int_id": 1,
            "lora_name": "policy_1",
            "lora_path": str(lora_path_1),
        }

        # Generate with updated adapter
        print("\nGenerating with updated (large/random) adapter...")
        responses_1 = generate_with_lora(base_url, prompts_formatted, lora_request_1)

        print("\nUpdated responses (lora_request_1):")
        for i, response in enumerate(responses_1):
            print(f"{i}: {response}")

        # Compare responses
        print("\n" + "=" * 60)
        print("COMPARISON:")
        print("=" * 60)

        responses_different = responses_0 != responses_1
        print(f"Responses are different: {responses_different}")

        for i, (r1, r2) in enumerate(zip(responses_0, responses_1)):
            print(f"\nPrompt {i}:")
            print(f"  lora_request_0:  {r1}")
            print(f"  lora_request_1:   {r2}")
            print(f"  Different: {r1 != r2}")

        if responses_different:
            print(
                "\n✅ SUCCESS: lora_request_1 update successfully changed model outputs!"
            )
        else:
            print("\n❌ FAILURE: lora_request_1 update did not change model outputs!")

        print("\nModifying LoRA parameters again.")
        for param in peft_model.parameters():
            if param.requires_grad:
                param.data = torch.zeros_like(param.data)

        lora_path_2 = work_dir / "lora_2"
        save_adapter_checkpoint(
            peft_model, lora_config, tensor_shapes, str(lora_path_2)
        )

        lora_request_2 = {
            "lora_int_id": 2,
            "lora_name": "policy_2",
            "lora_path": str(lora_path_2),
        }

        # Generate with updated adapter
        print("\nGenerating with updated (large/random) adapter...")
        responses_2 = generate_with_lora(base_url, prompts_formatted, lora_request_2)

        print("\nResponses 2:")
        for i, response in enumerate(responses_2):
            print(f"{i}: {response}")

        # Compare responses
        print("\n" + "=" * 60)
        print("COMPARISON:")
        print("=" * 60)

        responses_different = responses_1 != responses_2
        print(f"Responses are different: {responses_different}")

        for i, (r1, r2) in enumerate(zip(responses_1, responses_2)):
            print(f"\nPrompt {i}:")
            print(f"  lora_request_1:  {r1}")
            print(f"  lora_request_2:   {r2}")
            print(f"  Different: {r1 != r2}")

        if responses_different:
            print(
                "\n✅ SUCCESS: lora_request_2 update successfully changed model outputs!"
            )
        else:
            print("\n❌ FAILURE: lora_request_2 update did not change model outputs!")

    finally:
        # Cleanup
        print("\nShutting down VLLM server...")
        server_process.terminate()
        server_process.wait()


if __name__ == "__main__":
    main()

**Comment by hnt2601** [comment]
WIP
